pkg/networkpath/traceroute/tcp/utils_windows.go

```diff
@@ -0,0 +1,138 @@
+	"sync"
+	"time"
+
+	"golang.org/x/net/ipv4"
+	"golang.org/x/sys/windows"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/google/gopacket/layers"
+)
+
+
+// listenPackets takes in raw ICMP and TCP connections and listens for matching ICMP
+// and TCP responses based on the passed in trace information. If neither listener
+// receives a matching packet within the timeout, a blank response is returned.
+// Once a matching packet is received by a listener, it will cause the other listener
+// to be canceled, and data from the matching packet will be returned to the caller
+func (w *winrawsocket) listenPackets(timeout time.Duration, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, seqNum uint32) (net.IP, uint16, layers.ICMPv4TypeCode, time.Time, error) {
+	var icmpErr error
+	var wg sync.WaitGroup
+	var icmpIP net.IP
+	//var tcpIP net.IP
+	//var icmpCode layers.ICMPv4TypeCode
+	//var tcpFinished time.Time
+	var icmpFinished time.Time
+	var port uint16
+	wg.Add(1)
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	go func() {
+		defer wg.Done()
+		defer cancel()
+		icmpIP, _, _, icmpFinished, icmpErr = w.handlePackets(ctx, localIP, localPort, remoteIP, remotePort, seqNum)
+	}()
+	wg.Wait()
+
+	if icmpErr != nil {
+		_, icmpCanceled := icmpErr.(canceledError)
+		if icmpCanceled {
+			log.Trace("timed out waiting for responses")
+			return net.IP{}, 0, 0, time.Time{}, nil
+		}
+		if icmpErr != nil {
+			log.Errorf("ICMP listener error: %s", icmpErr.Error())
+		}
+
+		return net.IP{}, 0, 0, time.Time{}, fmt.Errorf("icmp error: %w", icmpErr)
+	}
+
+
+	// return the TCP response
+	return icmpIP, port, 0, icmpFinished, nil
+}
+
+
+// handlePackets in its current implementation should listen for the first matching
+// packet on the connection and then return. If no packet is received within the
+// timeout or if the listener is canceled, it should return a canceledError
+func (w* winrawsocket) handlePackets(ctx context.Context, localIP net.IP, localPort uint16, remoteIP net.IP, remotePort uint16, seqNum uint32) (net.IP, uint16, layers.ICMPv4TypeCode, time.Time, error) {
+	buf := make([]byte, 512)
+	tp := newTCPParser()
+	for {
+		select {
+		case <-ctx.Done():
+			return net.IP{}, 0, 0, time.Time{}, canceledError("listener canceled")
+		default:
+		}
+		
+		// the receive timeout is set to 100ms in the constructor, to match the
+		// linux side. This is a workaround for the lack of a deadline for sockets.
+		//err := conn.SetReadDeadline(now.Add(time.Millisecond * 100))
+		n, _, err := windows.Recvfrom(w.s, buf, 0)
+		if err != nil {
+			if err == windows.WSAETIMEDOUT {
+				continue;
+			}
+			if err == windows.WSAEMSGSIZE {
+				log.Warnf("Message too large for buffer")
+				continue
+			}
+			return nil, 0, 0, time.Time{}, err
+		}
+		log.Tracef("Got packet %+v", buf[:n])
+
+		if n < 20 { // min size of ipv4 header
+			continue
+		}
+		header, err := ipv4.ParseHeader(buf[:n])
```

Comment:
Worth noting ebpfless is using [gopacket's layer parsing](https://github.com/DataDog/datadog-agent/blob/2706272163c3131745d847dd1f534de245542c7d/pkg/network/tracer/connection/ebpfless_tracer.go#L115) to do something similar. It's not quite the same though, since we parse the TCP layer no matter what's in the IP layer (no early exit). If it's suitable, it could save some headache. It supports ICMP layers